### ISSUE TITLE
After seen all the messages , red dot is not disappearing from channel.

### ISSUE THREAD

**Issue by iosiappmemes** [GetStream/stream-chat-swift#1647]
Hi Team , I have found one bug while using StreamChat 4.0 SDK , 
I got a message from a friend , on the channel list it shows red dot on new message.  when I clicked on that channel and view all the messages and I came back from the messages screen , the red dot not disappear from the channel.

Here is the Screen shot for ref-
![Simulator Screen Shot - iPhone 13 Pro - 2021-11-25 at 13 10 51](https://user-images.githubusercontent.com/54673061/143399666-3eaa29e4-c801-43bb-85bd-4484967d5eef.png)

**Comment by adamrushy** [comment]
Hey, @iosiappmemes 

Thanks so much for sending this through to us, to help me investigate further can you provide the following:

- Steps to reproduce
- Environment details (iOS version, Xcode version etc...)
- Sample code

If you can reproduce this in a sample project or share with us your current implementation so I can see what's happening. I have tried reproducing in our Demo App but it's working as expected. 

Thank you,
Adam

**Comment by iosiappmemes** [comment]
@adamrushy , this is our client based live product so I can't share the project , and I don't have any sample project. 
Just let me know one thing, I am using default ui for chat , is there anything I need to call manually to make channel read all the messages ? any Boolean value or any api I need to call ?

**Comment by iosiappmemes** [comment]
IOS Version :- 15.0
Xcode Version: - 13.0

**Comment by iosiappmemes** [comment]
Steps to reproduce , 
Initiate user on two device , suppose names are A,B. 
A sent a message to B , when B goes back to channel list from messages controller , it is showing red dot, but B has already seen that message.
Channel should be refresh automatically and red dot should be disappear.

**Comment by b-onc** [comment]
@iosiappmemes can you share a video?

**Comment by vladpre92** [comment]
@iosiappmemes You should update to the last version. #1569 is the fix related to this problem and was added to version 4.2.0.

**Comment by iosiappmemes** [comment]
Okay Thanks.

**Comment by iosiappmemes** [comment]
Hi @vladpre92 , I reviewed your code and I have found the issue in your code .
Actually there are two issue , because of them red dot is not disappearing.
Even I have installed latest SDK as you mentioned 4.2.0, still there are issues available.

issue one :-  This is the code where thread is returning from without marking channel read.
Try to debug by break points.
guard channel?.latestMessages.first?.author.id != userId else {
            callback {
                completion?(nil)
            }
            return
        }
        
issue Second:- You are trying to mark channel read in this method which is not calling while opening the messages controller. This method only calls when we scroll the table view manually. so this is the second issue I have found.
 open func chatMessageListVC(_ vc: ChatMessageListVC, scrollViewDidScroll scrollView: UIScrollView) {
        if didReadAllMessages {
            channelController.markRead()
        }

**Comment by iosiappmemes** [comment]
I updated the code , I removed the first condition. 
and added this code in ViewDidAppear
if didReadAllMessages {
channelController.markRead()
}

Everything works fine now.

**Comment by iosiappmemes** [comment]
@vladpre92 , I am using  StreamChat (4.5.0) now.
But still am getting red dot issue. sharing a video , please check once.

https://user-images.githubusercontent.com/54673061/143849419-969a4ccf-b38a-447f-86c3-72613bce3415.mp4

**Comment by vladpre92** [comment]
@iosiappmemes Fix #1569 only solves the above mentioned use case:

`Initiate user on two device , suppose names are A,B.
A sent a message to B , when B goes back to channel list from messages controller , it is showing red dot, but B has already seen that message.`. (with the condition that user B already in messages controller when receiving the message.)

Also, I can't reproduce the behaviour presented in the video. In my case, channels are marked as read as expected when opening them, without the need to scroll.

**Comment by nuno-vieira** [comment]
Thanks a lot @vladpre92 for helping us out on this issue! Much appreciated! 

@iosiappmemes Unfortunately, our internal team can't reproduce this issue as well on the most recent versions.
I would double-check that the SDK version is being correctly updated and is not somehow cached. Which dependency manager are you using? 

If you are sure the SDK Version is being correctly updated, the only way we can help you is if you somehow send us your customizations so that we can pinpoint the problem. Otherwise, without reproducing it, we can't solve the issue. 

Thank you!

**Comment by iosiappmemes** [comment]
Okay @nuno-vieira , let me check once , will update you shortly .

**Comment by bardiab** [comment]
Hi we are also having this issue. Currently using version 4.2.0, double checked and the version is being updated correctly. 

User Flow:
`A sent B a message. B opens message and replies to A with a new message. B goes back to channel list from messages controller, showing dot with 1 new notification.`

**Comment by nuno-vieira** [comment]
Then, can you provide us with your `ChaChannelListVC` and `ChatChannelVC` code so that we can try and reproduce ourselves? This is not reproducible on our Sample Apps. Also, we have a lot of customers, and we don't have anyone reporting this issue.

Thanks

**Comment by konroj** [comment]
@nuno-vieira as @bardiab mentioned we're having same issue.
It's related to only one chat channel. It's not replicable. Even sending a new message doesn't count it down. Subclasses of ChatChannelListVC and ChatCannelVC doesn't change anything in case of logic and it works for (at least we for what we observed) most of the channels till now

**Comment by nuno-vieira** [comment]
@konroj 
> it's only related to one chat channel
Can you clarify? Which channel?

> It's not replicable 
What do you mean? So how can we replicate the issue ourselves so that we can solve the issue?

**Comment by nuno-vieira** [comment]
@konroj @bardiab @iosiappmemes As you can see from the video below, this is working as expected testing 4.5.0 on our sample app: 

https://user-images.githubusercontent.com/12814114/143948623-c8824440-3ec8-4495-81bd-bea8b228eeff.mp4

**Comment by konroj** [comment]
It's an active user with several channels.
The "stuck" unread message is only affecting one of his channels.
We are not able to replicate or debug this. But this issue exists.

Just got an info that someone has replied there and the unread message dot has gone after opening.

**Comment by nuno-vieira** [comment]
> It's an active user with several channels. The "stuck" unread message is only affecting one of his channels. We are not able to replicate or debug this. But this issue exists.
> 
> Just got an info that someone has replied there and the unread message dot has gone after opening.

Since the issue is gone, I'll close the issue. If you see it happen again, feel free to open the issue, but we need more details in order to fix it. I would recommend gathering information from the user and making sure that there's nothing wrong with the id of the user since the UI will only mark the message read if the message is not from the same user.

Thank you for reporting the issue anyways.

**Comment by nuno-vieira** [comment]
Hi

**Comment by nuno-vieira** [comment]
@iosiappmemes The only way I could see the issue coming from the SDK is that somehow in some iOS Version the `scroll view's event is not triggered. But if that is the case, you can override the `viewDidAppear` of the channel list and call: 
```swift
if didReadAllMessages {
     channelController.markRead()
}
```

with this, it still has the issue?

**Comment by iosiappmemes** [comment]
Hi @nuno-vieira , let me check this .

**Comment by iosiappmemes** [comment]
This can't be used because its implementation is private .

![Screen Shot 2021-11-30 at 3 51 10 PM](https://user-images.githubusercontent.com/54673061/144029438-639f97d2-221b-4af1-826b-eaccec266ef2.png)

**Comment by iosiappmemes** [comment]
Hi @nuno-vieira , 
I make changes in the code .
I just commented these few lines of code . it works fine for me. 
//        guard channel?.latestMessages.first?.author.id != userId else {
//            callback {
//                completion?(nil)
//            }
//            return
//        }

**Comment by nuno-vieira** [comment]
> This can't be used because its implementation is Internal .
> 
> ![Screen Shot 2021-11-30 at 3 51 10 PM](https://user-images.githubusercontent.com/54673061/144029438-639f97d2-221b-4af1-826b-eaccec266ef2.png)

I see.. we will make sure we change this, but for now, you can use this instead:
`messageListVC.listView.isLastCellFullyVisible`

**Comment by nuno-vieira** [comment]
> guard channel?.latestMessages.first?.author.id != userId

You can't comment those lines of codes. This is clear that something is wrong with your data. This only marks the channel read, if the last message is not from the current user.

**Comment by iosiappmemes** [comment]
Okay

**Comment by iosiappmemes** [comment]
> 

Let me try this one. will let you know.

**Comment by nuno-vieira** [comment]
@iosiappmemes Print out the user id of the current user (`userId`) and the `channel?.latestMessages.first?.author.id` and see if they are the same or if some of that id is `nil`.

**Comment by iosiappmemes** [comment]
@nuno-vieira , this is the actual issue, actually there are 5 members in the channel , and the last message is from my side (Logged in  user). So in this case this condition becoming wrong. so thread is returning from here. and red dot is staying there.
```
//        guard channel?.latestMessages.first?.author.id != userId else {
//            callback {
//                completion?(nil)
//            }
//            return
//        }
```

**Comment by iosiappmemes** [comment]
> Print out the user id of the current user (`userId`) and the `channel?.latestMessages.first?.author.id` and see if they are the same or if some of that id is `nil`.

Okay let me check.

**Comment by nuno-vieira** [comment]
> @nuno-vieira , this is the actual issue, actually there are 5 members in the channel , and the last message is from my side (Logged in user). So in this case this condition becoming wrong. so thread is returning from here. and red dot is staying there.
> 
> ```
> //        guard channel?.latestMessages.first?.author.id != userId else {
> //            callback {
> //                completion?(nil)
> //            }
> //            return
> //        }
> ```

If the last message if from your side (current user) how didn't you see it? If it was you sending the message? This is not making any sense

**Comment by iosiappmemes** [comment]
Yes , I don't know why it is showing red dot on the channel. the last message is from my side. and see I printed out the last message author id and current user id, Both are same.
last Message userID Optional("128")
My id ==> Optional("128")

**Comment by iosiappmemes** [comment]
> If the last message is from your side (current user) how didn't you see it? If it was you sending the message? This is not making any sense

Let me reproduce the issue , I will give you the complete details shortly.

**Comment by iosiappmemes** [comment]
> `messageListVC.listView.isLastCellFullyVisible`

This is the quick fix , it works fine now. thanks @nuno-vieira

**Comment by nuno-vieira** [comment]
> > `messageListVC.listView.isLastCellFullyVisible`
> 
> This is the quick fix , it works fine now. thanks @nuno-vieira This issue is related to scrolling methods, which are not calling while open the channel.

Nice 👍 We will make sure to add that on the `viewDidAppear` as well in future releases. For now, you can do that yourself as a workaround 👍

**Comment by vladpre92** [comment]
@nuno-vieira I think you could easily reproduce the problem with the following steps:
1. Downgrade to 4.1.0 or 4.0.0.
2. Reproduce the bug that #1569 fixes in 4.2.0 with the specification that after receiving the message from the other user (that makes unreadCount = 1), you send a message. This makes a channel with unreadCount = 1 and last message sent by current user. 
3. Kill the app or go back to channel list but without entering the unread channel again.
4. Update StreamChat to 4.2.0+.
5. You'll have a channel that will not be marked as read until the other user sends a message, bypassing `currentUserId != message.author.id` (that was introduced with 4.2.0). 

I think this is the issue actually but it will be solved automatically after new messages are sent in the channel. This problem occurred between migrations.

**Comment by iosiappmemes** [comment]
> I also think @iosiappmemes and @konroj problems are different. From @iosiappmemes's video sample you can see that the last message is not sent by current user.

@vladpre92 , That was the old case related to that video, 
But I got the another case as you can see I printed out the both the user ides 
> last Message userID Optional("128")
> My id ==> Optional("128")

**Comment by nuno-vieira** [comment]
@iosiappmemes @vladpre92 Yes we will change the implementation soon. We already have it planned. We will move the call to `viewDidAppear` as well as remove the `channel?.latestMessages.first?.author.id != userId` condition since it is a bit overprotective. With this in place, we should have all issues fixed and a more stable mark read functionality 👍 

We will update the ticket as soon we fix this.

Thank you guys for your efforts in reporting the issue 🙂

**Comment by bielikb** [comment]
Hi all,

we just released new [version 4.5.1](https://github.com/GetStream/stream-chat-swift/releases/tag/4.5.1) that contains the fix provided by @nuno-vieira.

Please try the new release and let us know if the changes provided has fixed your reported issues.

thanks!

Best,
Boris
